rts/motoko-rts/src/stabilization/layout.rs

```diff
@@ -0,0 +1,404 @@
+    Concat,
+    BigInt,
+    ObjInd,
+    Some,
+}
+
+const STABLE_TAG_ARRAY: u64 = 1;
+const STABLE_TAG_MUTBOX: u64 = 2;
+const STABLE_TAG_OBJECT: u64 = 3;
+const STABLE_TAG_BLOB: u64 = 4;
+const STABLE_TAG_BITS64: u64 = 5;
+const STABLE_TAG_REGION: u64 = 6;
+const STABLE_TAG_VARIANT: u64 = 7;
+const STABLE_TAG_CONCAT: u64 = 8;
+const STABLE_TAG_BIGINT: u64 = 9;
+const STABLE_TAG_OBJIND: u64 = 10;
+const STABLE_TAG_SOME: u64 = 11;
+
+#[repr(C)]
+pub struct StableTag(u64);
+
+impl StableObjectKind {
+    pub fn encode(&self) -> StableTag {
+        let value = match self {
+            StableObjectKind::Array => STABLE_TAG_ARRAY,
+            StableObjectKind::MutBox => STABLE_TAG_MUTBOX,
+            StableObjectKind::Object => STABLE_TAG_OBJECT,
+            StableObjectKind::Blob => STABLE_TAG_BLOB,
+            StableObjectKind::Bits64 => STABLE_TAG_BITS64,
+            StableObjectKind::Region => STABLE_TAG_REGION,
+            StableObjectKind::Variant => STABLE_TAG_VARIANT,
+            StableObjectKind::Concat => STABLE_TAG_CONCAT,
+            StableObjectKind::BigInt => STABLE_TAG_BIGINT,
+            StableObjectKind::ObjInd => STABLE_TAG_OBJIND,
+            StableObjectKind::Some => STABLE_TAG_SOME,
+        };
+        StableTag(value)
+    }
+}
+
+impl StableTag {
+    pub fn decode(&self) -> StableObjectKind {
+        match self.0 {
+            STABLE_TAG_ARRAY => StableObjectKind::Array,
+            STABLE_TAG_MUTBOX => StableObjectKind::MutBox,
+            STABLE_TAG_OBJECT => StableObjectKind::Object,
+            STABLE_TAG_BLOB => StableObjectKind::Blob,
+            STABLE_TAG_BITS64 => StableObjectKind::Bits64,
+            STABLE_TAG_REGION => StableObjectKind::Region,
+            STABLE_TAG_VARIANT => StableObjectKind::Variant,
+            STABLE_TAG_CONCAT => StableObjectKind::Concat,
+            STABLE_TAG_BIGINT => StableObjectKind::BigInt,
+            STABLE_TAG_OBJIND => StableObjectKind::ObjInd,
+            STABLE_TAG_SOME => StableObjectKind::Some,
+            _ => unsafe { rts_trap_with("Invalid tag") },
+        }
+    }
+}
+
+impl StableObjectKind {
+    fn deserialize(tag: Tag) -> StableObjectKind {
+        match tag {
+            // During the marking phase of the incremental GC, the mutator can see
+            // array slice information in the object tag.
+            TAG_ARRAY | TAG_ARRAY_SLICE_MIN.. => StableObjectKind::Array,
+            TAG_MUTBOX => StableObjectKind::MutBox,
+            TAG_OBJECT => StableObjectKind::Object,
+            TAG_BLOB => StableObjectKind::Blob,
+            TAG_BITS64 => StableObjectKind::Bits64,
+            TAG_REGION => StableObjectKind::Region,
+            TAG_VARIANT => StableObjectKind::Variant,
+            TAG_CONCAT => StableObjectKind::Concat,
+            TAG_BIGINT => StableObjectKind::BigInt,
+            TAG_OBJ_IND => StableObjectKind::ObjInd,
+            TAG_SOME => StableObjectKind::Some,
+            _ => unimplemented!("tag {tag}"),
```

Comment:
Is this actually unimplemented with more case to consider or impossible?

Reply:
Oh, it is indeed impossible. I need to improve this, by using `unreachable`